sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1173,8 +1173,20 @@ class Dataset[T] private[sql](
       joined = resolveSelfJoinCondition(joined)
     }
 
-    implicit val tuple2Encoder: Encoder[(T, U)] =
-      ExpressionEncoder.tuple(this.exprEnc, other.exprEnc)
+    // SPARK-37829: an outer-join requires the null semantics to represent missing keys.
+    // As we might be running on DataFrames, we need a custom encoder that will properly
+    // handle null top-level Rows.
+    def nullSafe[V](exprEnc: ExpressionEncoder[V]): ExpressionEncoder[V] = {
+      if (exprEnc.clsTag.runtimeClass != classOf[Row]) {
```

Comment:
This looks a bit ugly.

> I've tried simply wrapping CreateExternalRow with a null check and a number of tests started failing as they were assuming top-level rows couldn't be null.

Are they UT or end-to-end tests? If they are UT, we can simply update the tests because we have changed the assumption.